03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
 using namespace asset;
 using namespace video;
 
+// TODO[Przemek]: update comments
 
 // This is the most nuts thing you'll ever see, a header of HLSL included both in C++ and HLSL
 #include "app_resources/common.hlsl"
 
-
 // This time we create the device in the base class and also use a base class to give us an Asset Manager and an already mounted built-in resource archive
 class DeviceSelectionAndSharedSourcesApp final : public application_templates::MonoDeviceApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
 {
-		using device_base_t = application_templates::MonoDeviceApplication;
-		using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
-	public:
-		// Yay thanks to multiple inheritance we cannot forward ctors anymore
-		DeviceSelectionAndSharedSourcesApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
-			system::IApplicationFramework(_localInputCWD,_localOutputCWD,_sharedInputCWD,_sharedOutputCWD) {}
-
-		// we stuff all our work here because its a "single shot" app
-		bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+public:
+	// Yay thanks to multiple inheritance we cannot forward ctors anymore
+	DeviceSelectionAndSharedSourcesApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	// we stuff all our work here because its a "single shot" app
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Remember to call the base class initialization!
+		if (!device_base_t::onAppInitialized(std::move(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		// TODO: run-time sized buffers are not supporten in hlsl.. do it when testing glsl
```

Comment:
`ByteAddressBuffers` compile to equivalent of a runtime sized SSBO `GLSL` buffer with a single `uint32_t[]` block member